src/libs/TransactionUtils.ts

```diff
@@ -0,0 +1,386 @@
+        return format(createdDate, CONST.DATE.FNS_FORMAT_STRING);
+    }
+
+    return '';
+}
+
+function isDistanceRequest(transaction: Transaction): boolean {
+    const type = transaction?.comment?.type;
+    const customUnitName = transaction?.comment?.customUnit?.name;
+    return type === CONST.TRANSACTION.TYPE.CUSTOM_UNIT && customUnitName === CONST.CUSTOM_UNITS.NAME_DISTANCE;
+}
+
+function isReceiptBeingScanned(transaction: Transaction): boolean {
+    return [CONST.IOU.RECEIPT_STATE.SCANREADY, CONST.IOU.RECEIPT_STATE.SCANNING].some((value) => value === transaction.receipt.state);
+}
+
+/**
+ * Check if the transaction has a non-smartscanning receipt and is missing required fields
+ */
+function hasMissingSmartscanFields(transaction: Transaction): boolean {
+    return hasReceipt(transaction) && !isDistanceRequest(transaction) && !isReceiptBeingScanned(transaction) && areRequiredFieldsEmpty(transaction);
+}
+
+/**
+ * Check if the transaction has a defined route
+ */
+function hasRoute(transaction: Transaction): boolean {
+    return !!transaction?.routes?.route0?.geometry?.coordinates;
+}
+
+/**
+ * Get the transactions related to a report preview with receipts
+ * Get the details linked to the IOU reportAction
+ *
+ * @deprecated Use Onyx.connect() or withOnyx() instead
+ */
+function getLinkedTransaction(reportAction: ReportAction): Transaction | Record<string, never> {
+    let transactionID = '';
+
+    if (reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.IOU) {
+        transactionID = reportAction.originalMessage?.IOUTransactionID ?? '';
+    }
+
+    return allTransactions?.[`${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`] ?? {};
+}
+
+function getAllReportTransactions(reportID?: string): Transaction[] {
+    // `reportID` from the `/CreateDistanceRequest` endpoint return's number instead of string for created `transaction`.
+    // For reference, https://github.com/Expensify/App/pull/26536#issuecomment-1703573277.
+    // We will update this in a follow-up Issue. According to this comment: https://github.com/Expensify/App/pull/26536#issuecomment-1703591019.
+    const transactions: Transaction[] = Object.values(allTransactions ?? {}).filter((transaction): transaction is Transaction => transaction !== null);
+    return transactions.filter((transaction) => `${transaction.reportID}` === `${reportID}`);
+}
+
+/**
+ * Checks if a waypoint has a valid address
+ */
+function waypointHasValidAddress(waypoint: RecentWaypoint | null): boolean {
+    return !!waypoint?.address?.trim();
+}
+
+/**
+ * Converts the key of a waypoint to its index
+ */
+function getWaypointIndex(key: string): number {
+    return Number(key.replace('waypoint', ''));
+}
+
+/**
+ * Filters the waypoints which are valid and returns those
+ */
+function getValidWaypoints(waypoints: WaypointCollection, reArrangeIndexes = false): WaypointCollection {
+    const sortedIndexes = Object.keys(waypoints).map(getWaypointIndex).sort();
```

Comment:
Woops I think this introduced a regression or is at least part of the cause for this issue https://github.com/Expensify/App/issues/29621.

By default [Array.prototype.sort()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) does the following. "The default sort order is ascending, built upon converting the elements into strings, then comparing their sequences of UTF-16 code units values.". So therefore the waypoints get out of order when you have 10+ because the number indexes are sorted as strings.

I'm working on a fix, which should be easy for this piece.

Reply:
cc @cristipaval since you reviewed as well